metadata/metadata.go

```diff
@@ -1,47 +1,91 @@
 package metadata
 
-import "fmt"
+import (
+	"fmt"
+	"sync"
+)
 
 // Metadata is made as a standalone package to avoid import cycle:
 // influxd -> flux -> flux/interpreter -> flux/execute -> flux
-type Metadata map[string][]interface{}
+type Metadata struct {
+	// Metadata is passed as a value so we must store the mutex as a pointer to ensure it does not get copied
+	lock *sync.RWMutex
+	meta map[string][]interface{}
+}
+
+func NewMetadata() Metadata {
+	return Metadata{
+		lock: &sync.RWMutex{},
+		meta: make(map[string][]interface{}),
+	}
+}
 
 func (md Metadata) Add(key string, value interface{}) {
-	md[key] = append(md[key], value)
+	md.lock.Lock()
+	defer md.lock.Unlock()
+
+	md.meta[key] = append(md.meta[key], value)
 }
 
 func (md Metadata) AddAll(other Metadata) {
-	for key, values := range other {
-		md[key] = append(md[key], values...)
+	md.lock.Lock()
+	defer md.lock.Unlock()
+
+	other.lock.RLock()
+	defer other.lock.RUnlock()
+
+	for key, values := range other.meta {
+		md.meta[key] = append(md.meta[key], values...)
 	}
 }
 
 // Range will iterate over the Metadata. It will invoke the function for each
 // key/value pair. If there are multiple values for a single key, then this will
 // be called with the same key once for each value.
 func (md Metadata) Range(fn func(key string, value interface{}) bool) {
-	for key, values := range md {
+	md.RangeSlices(func(key string, values []interface{}) bool {
 		for _, value := range values {
 			if ok := fn(key, value); !ok {
-				return
+				return false
 			}
 		}
+		return true
+	})
+}
+
+func (md Metadata) RangeSlices(fn func(key string, values []interface{}) bool) {
```

Comment:
I don't know if there's a reasonable rule of thumb to follow, but it would be great if there was some suggestion for when this is important to use in the docstring.

Might be hopeless since I imagine the number one reason you'd seek this out is after a panic has already been reported :shrug: :woozy_face:

Reply:
I removed `RangeSlices` now after the refactor.

Reply:
Ah, yeah. The refactor looks like it explains the usage patterns better.